### PR TITLE
Allow teachers to hide all lessons in a unit with one click

### PR DESCRIPTION
**Pull request by rshipp**
Add two buttons to the top of the unit overview page (right above the first lesson tile) that have the option to "show all" and "hide all" with a tooltip to the right of it that says "Make all lessons in this Unit visible or hidden for your students".

image

The new buttons tie in to the existing logic for the toggle buttons down the rest of the page, and flip them dynamically.

Questions:

* Should I add a storybook file? I looked at it briefly and it seemed like it would take some work to figure out how to hook up all the redux stuff. Or should there be any other types of tests?
* Do I want the new buttons to show disabled when all the lessons are either hidden or visible? right now they feel a little low feedback when you click them
* I'm making one network call per lesson in the unit, which is probably not too many, but still seems less than ideal. Is it worth adding a bulk endpoint or something to the controller so I can do one call, and adding all the JS to handle that appropriately? It seems like that might increase the scope of this task by quite a bit.
* There's a warning in the console about "componentWillUpdate has been renamed, and is not recommended for use... Please update the following components: Connect(BulkLessonVisibilityToggle)" - is that something I should be concerned about here?

## Links

- jira ticket: [TEACH-16](https://codedotorg.atlassian.net/browse/TEACH-16)

## Testing story

* Manual testing, and one new basic unit test case.

### CODE REVIEW

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import {connect} from 'react-redux';
+import {getStore} from '../../../redux';
+import PropTypes from 'prop-types';
+import Button from '@cdo/apps/templates/Button';
+import FontAwesome from '../../../templates/FontAwesome';
+import ReactTooltip from 'react-tooltip';
+import {toggleHiddenLesson} from '@cdo/apps/code-studio/hiddenLessonRedux';
+import {unitCalendarLesson} from '../../../templates/progress/unitCalendarLessonShapes';
+import _ from 'lodash';
+import style from './bulk-lesson-visibility-toggle.module.scss';
```
Comment (rshipp) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: these are entirely unordered, and some use an absolute syntax while others use a relative syntax. are there any standards for this?

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+BulkLessonVisibilityToggle.propTypes = {
+  lessons: PropTypes.arrayOf(unitCalendarLesson),
+
+  // redux provided
+  sectionId: PropTypes.number,
+  scriptName: PropTypes.string.isRequired
+};
+
+function toggleHiddenLessons(scriptName, sectionId, lessons, hidden) {
```
Comment (rshipp) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: is it ok that this function is just floating around down here, or should it be somewhere else?

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  lessons.forEach(lesson => {
+    // For some reason, sectionId is a number here, and needs to be a string
+    // for the redux toggle stuff to work.
```
Comment (rshipp) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: is this related to the propTypes above? https://github.com/code-dot-org/code-dot-org/pull/48863/files#diff-df37c341b4e24afa4446ab42d92a84705737c3bc034cf85dfdcd51183bedc078R50. or is there something else I'm missing?

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+export default connect(
+  state => ({
+    sectionId: state.teacherSections.selectedSectionId,
+    scriptName: state.progress.scriptName
+  }) /*,
+  dispatch => ({
+    toggleHiddenLesson(unitName, sectionId, lessonId, hidden) {
+      dispatch(toggleHiddenLesson(unitName, sectionId, lessonId, hidden));
+    }
+  }) */
```
Comment (rshipp) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: i copied this from `ProgressLessonTeacherInfo.jsx`, but it seems to work fine without it? do I need this?

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        }
+      />
+      <span data-tip data-for={tooltipId}>
+        <FontAwesome icon="info-circle" className={style.infoTipIcon} />
+      </span>
+      <ReactTooltip id={tooltipId} effect="solid">
+        <p>
+          Make all lessons in this Unit visible or hidden for your students.
+        </p>
+      </ReactTooltip>
```
Comment (rshipp) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx:46]: `HelpTip` has a [Portal](https://www.npmjs.com/package/react-portal) here, should I use one too? i'm not sure what the benefit is

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx (hunk 1 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: We don't have any standards for order unfortunately :( 

For absolute vs relative syntax: I'll use relative if it's in the the files is in the same directory and otherwise use `@cdo/apps`. I'm not sure if that's just a personal preference though

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import Button from '@cdo/apps/templates/Button';
+import FontAwesome from '../../../templates/FontAwesome';
+import ReactTooltip from 'react-tooltip';
+import {toggleHiddenLesson} from '@cdo/apps/code-studio/hiddenLessonRedux';
+import {unitCalendarLesson} from '../../../templates/progress/unitCalendarLessonShapes';
+import _ from 'lodash';
+import style from './bulk-lesson-visibility-toggle.module.scss';
+import i18n from '@cdo/locale';
+
+function BulkLessonVisibilityToggle({lessons, sectionId, scriptName}) {
```
Comment (bethanyaconnor) [inline apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: if it's not too much trouble, it would be great to use `unitName` instead of `scriptName`. We're in the process of change the term "script" to "unit"

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx (hunk 2 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: I normally put these at the top, but, again, not sure if that's just personal preference

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx (hunk 5 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx:46]: Out of curiosity, why didn't you use `HelpTip` here? (I'm not sure what benefit a Portal gets you either though)

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx (hunk 4 above)
Comment (bethanyaconnor) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx]: I would remove the commented out code here. Not sure why it's in `ProgressLessonTeacherInfo`

**Review by bethanyaconnor — APPROVED** [review]
A couple of questions but looks good!

File: apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx (hunk 5 above)
Comment (rshipp) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx:46]: It just doesn't look very good, and doesn't have a way to pass in options to override the default look. I suppose I also could have modified it to let me pass in classNames/icons stuff, would that make more sense?

Comment (bethanyaconnor) [reply apps/src/code-studio/components/progress/BulkLessonVisibilityToggle.jsx:46]: I think this is fine in that case! Was just curious

**Review by hannahbergam — APPROVED** [review]
The question you ask about how it looks a little low feedback without disabling whichever button was just pressed- do the 'visible/hidden' fields in each lesson update automatically, or does it require a page load?

**Comment by rshipp** [comment]
The visible/hidden toggle for each lesson updates immediately, without a page load. They're just lower on the screen, so depending on the screen size/etc it might not be visible, or immediately clear that something happened.
